File: javascript/leetcode/230. Kth Smallest Element in a BST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
let kth, res;

var kthSmallest = function(root, k) {
  kth = k;
  res = 0;
  small(root);
  return res;
};

function small(root) {
  if (!root || kth === 0) return;
  small(root.left);
  kth--;
  if (kth === 0) res = root.val;
  else small(root.right);
}

File: go/leetcode/230. Kth Smallest Element in a BST.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func kthSmallest(root *TreeNode, k int) int {     
    kth := k 
    res := 0
    small(root, &res, &kth)
    return res
}

func small(root *TreeNode, res, kth *int) {
    if root == nil || *kth == 0 {
        return
    }
    
    small(root.Left, res, kth)
    *kth--
    if *kth == 0 {
        *res = root.Val
    } else {
        small(root.Right, res, kth)
    }
}